include/emds/emds.h

```diff
@@ -0,0 +1,189 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+/** @file
+ *  @defgroup emds Emergency Data Storage
+ *  @ingroup emds
+ *  @{
+ *  @brief Emergency Data Storage API
+ */
+
+#ifndef EMDS_H__
+#define EMDS_H__
+
+#include <stddef.h>
+#include <sys/types.h>
+#include <sys/util.h>
+#include <sys/slist.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @struct emds_entry
+ *
+ * Information about entries used to store data in the emergency data storage.
+ */
+struct emds_entry {
+	/** Unique id for each static and dynamic entry. */
+	uint16_t id;
+	/** Pointer to data that shall be stored. */
+	uint8_t *data;
+	/** Length of data that shall be stored. */
+	size_t len;
+}
+
+/**
+ * @struct emds_dynamic_entry
+ *
+ * Entries with a node element, used for dynamic entries. These are registered
+ * using a call to emds_entry_add().
+ */
+struct emds_dynamic_entry {
+	struct emds_entry entry;
+	sys_snode_t node;
+}
+
+/**
+ * @brief Define a static entry for emergency data storage items.
+ *
+ * @param name The entry name.
+ * @param _id Unique id for entry. This value and not overlap with any other
+ *            value.
+ * @param _data Data pointer to be stored at emergency data store.
+ * @param _len Length of data to be stored at emergency data store.
+ *
+ * This creates a variable _name prepended by emds_.
+ */
+#define EMDS_STATIC_ENTRY_DEFINE(_name, _id, _data, _len)                      \
+	static const STRUCT_SECTION_ITERABLE(emds_entry, emds_##_name) = {     \
+		.id = _id,                                                     \
+		.data = (uint8_t *)_data,                                      \
+		.len = _len,                                                   \
+	}
+
+/**
+ * @brief Initialize the emergency data storage.
+ *
+ * Initializes the emergency data storage. This need to called before starting
+ * to add entries to the emds_dynamic_entry and doing restore.
+ *
+ * @retval 0 Success
+ * @retval -ERRNO errno code if error
+ */
+int emds_init(void);
+
+/**
+ *  @brief Add entry that should be saved/restored when emergency data storage is called.
+ *
+ * This will add the entry to the dynamic entry list, and will when the
+ * @ref 'emds_store' function is called take the given data pointer and store
+ * the data to the emergency data storage.
+ *
+ * @param entry Entry to add to list.
+ *
+ * @retval 0 Success
+ * @retval -ERRNO errno code if error
+ */
+int emds_entry_add(struct emds_dynamic_entry *entry);
+
+/**
+ * @brief Start the emergency data storage process.
+ *
+ * This will trigger the process of storing all data that has been regitster to
+ * be stored. When it is finished with storing the data the data should not be
+ * changed and the board should be halted or rebooted at the end. This is a time
+ * critical process and will be processed as fast as possible. This will store
+ * all data register either though the 'emds_entry_add' or 'EMDS_ENTRY_DEFINE'.
```

Comment:
```suggestion
 * Triggers the process of storing all data registered to be stored.
 * All data registered either through 'emds_entry_add' function or the 
 * 'EMDS_ENTRY_DEFINE' macro is stored.
 * Once the data storage is completed, the data should not be
 * changed, and the board should be halted or rebooted at the end.
 * This is a time-critical process and will be processed as fast as possible.
```
Also, should 'emds_entry_add' and 'EMDS_ENTRY_DEFINE' be replaced by @ref emds_entry_add and @ref EMDS_ENTRY_DEFINE? And I only see EMDS_STATIC_ENTRY_DEFINE... Could it be that one you are referencing?